nav2_collision_monitor/include/nav2_collision_monitor/polygon_source.hpp

```diff
@@ -0,0 +1,108 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef NAV2_COLLISION_MONITOR__POLYGON_SOURCE_HPP_
+#define NAV2_COLLISION_MONITOR__POLYGON_SOURCE_HPP_
+
+#include <memory>
+#include <vector>
+#include <string>
+
+#include "nav2_msgs/msg/polygons_array.hpp"
+
+#include "nav2_collision_monitor/source.hpp"
+
+namespace nav2_collision_monitor
+{
+
+/**
+ * @brief Implementation for polygon source
+ */
+class PolygonSource : public Source
+{
+public:
+  /**
+   * @brief PolygonSource constructor
+   * @param node Collision Monitor node pointer
+   * @param source_name Name of data source
+   * @param tf_buffer Shared pointer to a TF buffer
+   * @param base_frame_id Robot base frame ID. The output data will be transformed into this frame.
+   * @param global_frame_id Global frame ID for correct transform calculation
+   * @param transform_tolerance Transform tolerance
+   * @param source_timeout Maximum time interval in which data is considered valid
+   * @param base_shift_correction Whether to correct source data towards to base frame movement,
+   * considering the difference between current time and latest source time
+   */
+  PolygonSource(
+    const nav2_util::LifecycleNode::WeakPtr & node,
+    const std::string & source_name,
+    const std::shared_ptr<tf2_ros::Buffer> tf_buffer,
+    const std::string & base_frame_id,
+    const std::string & global_frame_id,
+    const tf2::Duration & transform_tolerance,
+    const rclcpp::Duration & source_timeout,
+    const bool base_shift_correction);
+  /**
+   * @brief PolygonSource destructor
+   */
+  ~PolygonSource();
+
+  /**
+   * @brief Data source configuration routine. Obtains ROS-parameters
+   * and creates subscriber.
+   */
+  void configure();
+
+  /**
+   * @brief Adds latest data from polygon source to the data array.
+   * @param curr_time Current node time for data interpolation
+   * @param data Array where the data from source to be added.
+   * Added data is transformed to base_frame_id_ coordinate system at curr_time.
+   */
+  void getData(
+    const rclcpp::Time & curr_time,
+    std::vector<Point> & data) const;
+
+  void convertPolygonStampedToPoints(
+    const geometry_msgs::msg::PolygonStamped & polygon,
+    std::vector<Point> & data) const;
+
+protected:
+  /**
+   * @brief Getting sensor-specific ROS-parameters
+   * @param source_topic Output name of source subscription topic
+   */
+  void getParameters(std::string & source_topic);
+
+  /**
+   * @brief PolygonSource data callback
+   * @param msg Shared pointer to PolygonSource message
+   */
+  void dataCallback(nav2_msgs::msg::PolygonsArray::ConstSharedPtr msg);
+
+  // ----- Variables -----
+
+  /// @brief PolygonSource data subscriber
+  rclcpp::Subscription<nav2_msgs::msg::PolygonsArray>::SharedPtr data_sub_;
+
+  /// @brief Latest data obtained
+  nav2_msgs::msg::PolygonsArray::ConstSharedPtr data_;
```

Comment:
This should use standard messages, not custom ones. You can have it store a vector of data and then clear the vector after its been populated by the monitor

Reply:
hmm, I see one challenge, how would you not store duplicates in the vector? By cross-checking every field of the msg::Polygon against the Polygons already in the vector and not adding it if it matches?

Reply:
Will these polygons have headers? Perhaps keep the latest of each header typed? Maybe as a parameter so that others might want to use all

Reply:
They will need to have headers to check individually if they are too old to be valid. What do yoh mean with keep it "typed", can you clarify?

Reply:
I mean if you have N polygons, check the header values and keep only the newest of each reference frame, but as a parameterized option

Reply:
What would you propose?

Reply:
I stated the only alternatives I have already, either custom message or tracking with std::set + custom comparator as mentioned above. The custom msg seems simpler but I understand wanting to stick to standard messages...

Reply:
OK, lets go with that then. I'm not sure a custom comparator is needed, I believe `msg == msg` is implemented for all ROS message structs.

Reply:
I implemented the logic that will store incoming polygons to a vector plus some additional logic to keep this vector up to date:

- if a new polygon arrives that is equivalent to an existent one in the vector but is more recent, then it replaces it
- if any polygon is older than source_timeout it's cleared from the vector

The problem I'm facing is that if we're publishing a "moving polygon", then there will be a trail (of the old poses of this polygon) because the old poses in the vector do not satisfy any of the above stated conditions initially. Only after they are older than source_timeout will they be cleared

![image](https://github.com/ros-planning/navigation2/assets/14879660/ea68e40e-c74a-464a-a1ad-ecb40b664a04)

Reply:
That's what I thought about the frame_id containing some unique element of the particular polygon - but I could see how that wouldn't work if everything is pre-transformed. I'd really ideally like to **not** create a new collision-monitor specific message type, since that wouldn't be portable & it would require a shim to any natural messages they might use. A name or uuid would do the trick, but not having something like that in the msg definition causes a problem. 

I suppose another option is to have `N` polygon sources each of their own topic for each individual polygon. For these polygons you're using, how many are we talking about would be of interest? 1? 2?  10? 100? Would having 1 topic-source per robot footprint be doable?

Another option: if the polygons are stamped and we make some statement that the polygons to run multiple in 1 instance need to be sychronized, we could check for any that have a timestamp older than the latest by `eta` and remove those. Think something like a depth processing pipeline with stereo image pairs. Its a little hacky. 

The last option is the first you had; just have a custom message. I suppose we could do that, though I'd probably want that to include the polygon, stamp, and some unique identifier so we can fully solve the problem if we had that. The array might not be necessary at that point then to allow it to be more distributed. Though, I suppose it could continue to be a vector as long as we don't just store the last message but buffer `N` data sources

Reply:
I implemented my proposal and it works quite well, I tried to minimize the risk of wrongfully clearing polygons by making `arePolygonsSimilar` check 2 things, the polygons are considered similar if:

1- they have the same number of points
2- Pair by pair, their respective points all have a distance smaller than `polygon_similarity_threshold_`

Reply:
@smac is the proposes solution satisying to you? (I'll worry abou the testing at the end to avoid double work)